### PR TITLE
upgrade workflow-api to list (un)stashed files

### PR DESCRIPTION
**Pull request by joseblas**

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java [hunk 1]
```diff
@@ -111,29 +140,53 @@ public boolean isAllowEmpty() {
...
+                        logger.println("stashed file " + file.getName());
+                    }
+
+                });
+            }
+        }
+
+        @Override
+        protected Void run() throws Exception {
+            List<File> stash = StashManager.stash(getContext().get(Run.class), step.name, getContext().get(FilePath.class), getContext().get(Launcher.class), getContext().get(EnvVars.class), getContext().get(TaskListener.class), step.includes, step.excludes,
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:162]: `File`? Does not smell right. That could only refer to paths on the controller, and stashes are archived into a tarball. If these “files” are on an agent, then you cannot use `File` to refer to them. You _could_ use `FilePath` but (as above) you probably just meant to use `String` (relative paths).

File: pom.xml [hunk 2]
```diff
@@ -64,7 +64,7 @@
     <properties>
         <revision>2.23</revision>
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.249.1</jenkins.version>
+        <jenkins.version>2.264-SNAPSHOT</jenkins.version>
```
[the remark this answers is not in the document]
Comment (joseblas) [reply pom.xml:67]: ok about removing reflection, but some changes will needed in Jenkins-core (https://github.com/jenkinsci/jenkins/pull/5023#pullrequestreview-514587286) so a new version will be needed.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java [hunk 3]
```diff
@@ -68,35 +83,49 @@ public String getIncludes() {
...
         this.allowEmpty = allowEmpty;
     }
 
-    @Override public StepExecution start(StepContext context) throws Exception {
+    public boolean isVerbose() {
+        return verbose;
+    }
+
+    @DataBoundSetter
+    public void setVerbose(boolean verbose) {
```
[the remark this answers is not in the document]
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: happy to discuss it

File: pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jglick) [reply pom.xml:67]: Preparing this in #134.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: indeed, this is a good example why printing should be done in here, if the list is truncated or not is responsibility of this plugin, not workflow-api which has to supply that list only.

[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: If you just get rid of the step option and print some reasonable level of detail by default then you need no API change and no patch to this plugin.

**Comment by jglick** [comment]
See #54.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java (hunk 1 above)
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:162]: my goal is to pass something neutral, like a file (or FilePAth), or even a String. I did passed a File in order to provide more info in higher levels, not meaning to pass the actual file. Indeed, something like a DTO, which could be created.

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:162]: Just `String` should suffice. `stash` and `unstash` deal with paths relative to the contextual working dir, so the absolute file path is not really interesting.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: I'd rather do the right thing, which is (to me)to pass the files (un)stashed back to the step to print it if appropriate.

[the remark this answers is not in the document]
Comment (rpau) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: The verbose parameter needs to be optional as the task describes, and by default should be false to ensure that the behavior is consistent with the previous releases. So, this does not need to be trunked.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java (hunk 1 above)
Comment (rpau) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:162]: I would say an URI :)

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: Why do we care if the behavior is identical to previous releases? This is just diagnostic logging as a convenience.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java [hunk 4]
```diff
@@ -111,29 +140,53 @@ public boolean isAllowEmpty() {
...
-            StashManager.stash(getContext().get(Run.class), step.name, getContext().get(FilePath.class), getContext().get(Launcher.class), getContext().get(EnvVars.class), getContext().get(TaskListener.class), step.includes, step.excludes,
+        private void verbose(List<File> files) throws Exception {
+            if (step.verbose) {
+                files.stream().forEach(file -> {
+                    PrintStream logger = null;
+                    try {
+                        String name = getContext().get(Run.class).getParent().getName();
+                        logger = getContext().get(TaskListener.class).getLogger();
+                        String gibberish = file.getCanonicalPath().split("/" + name + "/")[0];
+                        logger.println("stashed file " + StringUtils.substring(file.getCanonicalPath(), gibberish.length()));
```
[the remark this answers is not in the document]
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:151]: how?

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: for is more about responsibilities, and for me there is no need to log as a convenience. The log has to be printed in the StashStep if verbose. And StashManager to provide a list a changes.

[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/support/steps/stash/StashStep.java:123]: At least pass the `verbose` flag into the `StashManager` methods and let the implementation handle the printing.

**Comment by jglick** [comment]
Obsolete I guess?

**Pull request closed without merging**